clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -4838,35 +4846,51 @@ void CodeGenFunction::EmitTrapCheck(llvm::Value *Checked,
+      llvm::SmallVector<llvm::Value *, 1> RuntimeCallArgs;
       // Emit calls to one of the interfaces defined in
       // `bounds_safety_soft_traps.h`
       switch (CGM.getCodeGenOpts().getBoundsSafetyTrapMode()) {
-      case CodeGenOptions::BoundsSafetyTrapModeKind::SoftCallWithTrapCode: {
-        // void __bounds_safety_soft_trap_c(uint16_t);
-        // TODO: Set correct value from the `TrapReason` object
-        // (rdar://162824128).
-        SoftTrapCallArg = llvm::ConstantInt::get(CGM.Int16Ty, 0);
+      case CodeGenOptions::BoundsSafetyTrapModeKind::SoftCallMinimal: {
+        // void __bounds_safety_soft_trap();
+        FnType = llvm::FunctionType::get(CGM.VoidTy, {}, false);
+        assert(RuntimeCallArgs.empty());
         break;
       }
       case CodeGenOptions::BoundsSafetyTrapModeKind::SoftCallWithTrapString: {
         // void __bounds_safety_soft_trap_s(const char*);
-        if (!TrapMessage.empty()) {
-          SoftTrapCallArg =
-              CGM.GetOrCreateGlobalStr(TrapMessage, Builder, "trap.reason");
-        } else {
-          SoftTrapCallArg = llvm::Constant::getNullValue(CGM.Int8PtrTy);
-        }
+        FnType = llvm::FunctionType::get(CGM.VoidTy, {CGM.Int8PtrTy}, false);
+        if (!TrapMessage.empty())
+          RuntimeCallArgs.push_back(
+              CGM.GetOrCreateGlobalStr(TrapMessage, Builder, "trap.reason"));
+        else
+          RuntimeCallArgs.push_back(
+              llvm::Constant::getNullValue(CGM.Int8PtrTy));
         break;
       }
       default:
         llvm_unreachable("Unhandled BoundsSafetyTrapMode");
       }
-      FnType = llvm::FunctionType::get(CGM.VoidTy, {SoftTrapCallArg->getType()},
-                                       false);
       auto TrapFunc = CGM.CreateRuntimeFunction(
           FnType, CGM.getCodeGenOpts().BoundsSafetySoftTrapFuncName);
-      TrapCall = EmitNounwindRuntimeCall(TrapFunc, {SoftTrapCallArg});
+
+      // Change calling convention if necessary. This needs to be kept in sync
+      // with `bounds_safety_soft_traps.h`.
+      if (ClBoundsSafetySoftTrapPreserveAllCC &&
+          (CGM.getTarget().getTriple().isAArch64() ||
+           CGM.getTarget().getTriple().isX86_64())) {
+        auto *Fn = cast<llvm::Function>(TrapFunc.getCallee());
+        // This calling convention is used to reduce the codesize at the
+        // callsite. It makes more registers be callee saved which means if
+        // those registers are being used in the caller they no longer need to
+        // be spilled. Unfortunately despite the name this calling convention
+        // doesn't preserve all registers. We should add a new calling
+        // convention that actually preserves almost all the general purposes
+        // registers (rdar://164341162).
+        Fn->setCallingConv(llvm::CallingConv::PreserveAll);
+      }
+
+      TrapCall = EmitNounwindRuntimeCall(TrapFunc, RuntimeCallArgs);
+      TrapCall->addFnAttr(llvm::Attribute::NoMerge);
```

Comment:
@fhahn Should we be putting any the `memory(...)` attributes on the soft trap runtime function?

Implementations of the soft trap runtime are supposed to log the event (in whatever way they see fit) and return. While a soft trap runtime may maintain its own global state that is not supposed to be visible to the rest of the program.

I ask because I wonder if code would be better optimized (e.g. avoid spilling recently modified values in the caller  that are currently in registers back to memory just in case the called function would read it) if there were attributes communicating how the soft trap runtime function uses memory.